evm-validate/src/lib.rs

```diff
@@ -417,3 +468,764 @@ fn validate_transaction_hashes(
+            u256.try_into().unwrap()
+        });
+        let max_fee_per_gas: Option<u128> = tx_max_fee_per_gas_opt.map(|bytes| {
+            let u256 = U256::try_from_be_slice(bytes).expect("invalid max fee per gas");
+            u256.try_into().unwrap()
+        });
+        let chain_id = tx_chain_id_opt; 
+        let cumulative_gas_used = tx_cumulative_gas_used_opt.unwrap();
+        let contract_address: Option<Address> = tx_contract_address_opt.map(|a| a.try_into().unwrap());
+        let logs_bloom = tx_logs_bloom_opt.unwrap();
+        let tx_type = tx_type_opt.unwrap();
+        let status = tx_status_opt.unwrap();
+        let expected_sighash = tx_sighash_opt;
+        let access_list: Option<AccessListWrapper> = tx_access_list_opt.map(|bytes| bincode::deserialize(bytes).unwrap_or_else(|_|{
+            println!("failed to deserialize access list, block {}, tx_hash {}", block_num, expected_hash);
+            return AccessListWrapper(vec![]);
+        }));
+        let access_list: Option<AccessList> = access_list.map(|list| list.into());
+        let max_fee_per_blob_gas: Option<u128> = tx_max_fee_per_blob_gas_opt.map(|bytes| {
+            let u256 = U256::try_from_be_slice(bytes).expect("invalid max fee per blob gas");
+            u256.try_into().unwrap()
+        });
+        let blob_versioned_hashes: Option<Vec<FixedBytes<32>>> = tx_blob_versioned_hashes_opt.map(|bytes| {
+            bytes.chunks(32).map(|chunk| FixedBytes::from_slice(chunk)).collect()
+        });
+        
+        // if the block num has changed, store the previous tx receipts and tx envelopes in the mapping, clear the receipts vec and update the current block num
+        if block_num != current_block_num {
+            if !block_tx_receipts.is_empty() {
+                let receipt_root = calculate_receipt_root(&block_tx_receipts);
+                let transactions_root = calculate_transaction_root(&block_tx_envelopes);
+                receipts_root_by_block_num_mapping.insert(current_block_num, receipt_root);
+                transactions_root_by_block_num_mapping.insert(current_block_num, transactions_root);
+                block_tx_receipts.clear();
+                block_tx_envelopes.clear();
+            }
+            current_block_num = block_num;
+        }
+
+        // validate sighash
+        match expected_sighash {
+            Some(expected_sighash) => {
+                let sighash: [u8; 4] = input[..4].try_into().expect("input must be at least 4 bytes long for a tx with a sighash");
+                if sighash != expected_sighash {
+                    println!("sighash mismatch at block {}, tx_idx {}.\nExpected:\n{:?},\nFound:\n{:?}", block_num, tx_idx, expected_sighash, sighash);
+                }
+            },
+            None => {
+                if input.len() > 0 {
+                    println!("sighash is None, How to know if this tx has UTF-8 input?");
```

Comment:
I don't know exactly what to do in this case.
A simple ETH transfer (no code execution), but the input might not be empty because the sender is sending text. ie: https://etherscan.io/tx/0xeaac25cad195c65c7e84dcc716e0b92d9f875435e93dab329e566e7900cdc2ac
In the above case, I think the sighash should be empty, but in Envio's case, it's not. It's taking the 4bytes anyway.
I don't think there is a way for Envio to differentiate this tx.
I think the best approach here is to do nothing.

Reply:
sighash is the first four bytes by definition afaik so it should take the first four bytes regardless of the meaning.

there is something wrong if the case in 704 happens in practice, can put an error there